File: packages/provide-uterm-go/colors/mode.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Unified color-mode dispatchers — string and byte-slice variants.

package colors

// ColorMode selects the color-mode filter applied by ApplyColorMode.
type ColorMode string

// Supported color modes.
const (
	// ModePassthrough returns data unchanged (zero-cost hot path).
	ModePassthrough ColorMode = "passthrough"
	// Mode256 downgrades truecolor SGR to the xterm-256 cube.
	Mode256 ColorMode = "256"
	// Mode16 downgrades truecolor SGR to the base 16-color palette.
	Mode16 ColorMode = "16"
)

// ApplyColorMode applies a color-mode filter to ANSI text containing SGR
// sequences. Modes other than ModePassthrough and Mode256 downgrade to the
// base 16-color palette (mirroring the Python Literal contract).
func ApplyColorMode(data string, mode ColorMode) string {
	if mode == ModePassthrough {
		return data
	}
	if mode == Mode256 {
		return DowngradeTo256(data)
	}
	return DowngradeTo16(data)
}

// ApplyColorModeBytes applies a color-mode filter to raw bytes containing
// SGR sequences.
//
// The Python implementation decodes bytes as latin-1, runs the SGR regex,
// and re-encodes. Latin-1 maps bytes 0-255 one-to-one onto code points
// 0-255 and the SGR pattern is pure ASCII, so running the regex directly
// over the bytes is byte-for-byte equivalent.
func ApplyColorModeBytes(data []byte, mode ColorMode) []byte {
	if mode == ModePassthrough {
		return data
	}
	target := Mode16
	if mode == Mode256 {
		target = Mode256
	}
	return SGRRegexp.ReplaceAllFunc(data, func(m []byte) []byte {
		return []byte(RewriteParams(string(m[2:len(m)-1]), target))
	})
}

File: packages/provide-uterm-ts/src/colors/mode.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Unified color-mode dispatcher — handles strings or bytes transparently.
 *
 * Port of the Python module `provide.uterm.colors.mode`.
 */

import { downgradeTo16, downgradeTo256 } from "./downgrade.ts";

/** Color-mode filter selector. */
export type ColorMode = "passthrough" | "256" | "16";

/**
 * Apply a color-mode filter to ANSI text containing SGR sequences.
 *
 * Modes other than `"passthrough"` and `"256"` downgrade to the base
 * 16-color palette, mirroring the Python `Literal` contract.
 */
export function applyColorMode(data: string, mode: ColorMode): string {
  if (mode === "passthrough") {
    return data;
  }
  return mode === "256" ? downgradeTo256(data) : downgradeTo16(data);
}

/**
 * Apply a color-mode filter to raw bytes containing SGR sequences.
 *
 * The Python implementation decodes bytes as latin-1, runs the SGR regex,
 * and re-encodes. Latin-1 maps bytes 0-255 one-to-one onto code points 0-255
 * and both the SGR pattern and every replacement it emits are pure ASCII, so
 * the round-trip is byte-for-byte faithful.
 */
export function applyColorModeBytes(data: Uint8Array, mode: ColorMode): Uint8Array {
  if (mode === "passthrough") {
    return data;
  }
  const text = Buffer.from(data).toString("latin1");
  const filtered = mode === "256" ? downgradeTo256(text) : downgradeTo16(text);
  return new Uint8Array(Buffer.from(filtered, "latin1"));
}
